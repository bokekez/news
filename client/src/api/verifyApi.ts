export const verifyUser = async (token: string): Promise<string> => {
  const response = await fetch('http://localhost:8000/login/verify', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ token }),
  });

  if (!response.ok) {
    const errorData = await response.json();
    throw new Error(errorData.error);
  }

  const data = await response.json();
  return data;
};