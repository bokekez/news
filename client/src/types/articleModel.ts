export interface Article {
  source: {
    id: string;
    name: string;
  };
  author: string;
  title: string;
  description: string;
  url: string;
  urlToImage: string;
  publishedAt: string;
  content: string;
  id?: string;
}

export interface ArticleCards extends Partial<Article> {
  widget?: boolean;
}

export interface ArticleResposne {
  article: Article;
  id?: string;
}