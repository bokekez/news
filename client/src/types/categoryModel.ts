import { ReactNode } from 'react';

export interface CategoryList {
  name: string;
  img: ReactNode;
}