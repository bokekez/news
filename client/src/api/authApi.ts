import { LoginResponse, RegisterPayload, RegisterResponse } from "../types/loginModel";
import { MESSAGES } from "../constants/Messages";
const BASE_URL = 'http://localhost:8000';

export const loginUser = async (username: string, password: string): Promise<LoginResponse> => {
  try {
    const response = await fetch(`${BASE_URL}/login`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ username, password }),
    });

    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(errorData.error);
    }

    const {data} = await response.json();
    return data
  } catch (error: unknown) {
    throw new Error(error instanceof Error ? error.message : MESSAGES.ERROR.UNKNOWN);
  }
};

export const registerUser = async (payload: RegisterPayload): Promise<RegisterResponse> => {
  console.log(payload)
  try {
    const response = await fetch(`${BASE_URL}/login/register`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(errorData.error);
    }

    return await response.json();
  } catch (error: unknown) {
    throw new Error(error instanceof Error ? error.message : MESSAGES.ERROR.UNKNOWN);
  }
};
