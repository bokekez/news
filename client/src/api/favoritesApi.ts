import { Article, ArticleResposne } from '../types/articleModel';
import { MESSAGES } from '../constants/Messages';
const BASE_URL = 'http://localhost:8000/favorites/';

export const fetchFavorites = async (userId: string): Promise<Article> => {
  try {
    const response = await fetch(`${BASE_URL}?userId=${userId}`);
    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error || 'Failed to fetch favorites');
    }
    const data = await response.json();
    data.favorites.forEach((fav: ArticleResposne) => {
      if (fav.article) {
        fav.article.id = fav.id;
      }
    });
    return data.favorites.map((fav: ArticleResposne) => fav.article);
  } catch (error: unknown) {
    throw new Error(error instanceof Error ? error.message : MESSAGES.ERROR.UNKNOWN);
  }
};

export const saveFavorite = async (userId: number, article: Article): Promise<Article[]> => {
  try {
    const response = await fetch(BASE_URL, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ userId, article }),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error || 'Failed to save favorite');
    }

    const data = await response.json();
    return data.favorite;
  } catch (error: unknown) {
    throw new Error(error instanceof Error ? error.message : MESSAGES.ERROR.UNKNOWN);
  }
};

export const deleteFavorite = async (favoriteId: number): Promise<string> => {
  try {
    const response = await fetch(`${BASE_URL}${favoriteId}`, { method: 'DELETE' });
    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error || 'Failed to delete favorite');
    }
    const data = await response.json();
    return data.message;
  } catch (error: unknown) {
    throw new Error(error instanceof Error ? error.message : MESSAGES.ERROR.UNKNOWN);
  }
};
