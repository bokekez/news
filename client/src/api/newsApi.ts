import { Article } from '../types/articleModel';
const BASE_URL = 'http://localhost:8000/news/';

export const fetchNewsBySearchTerm = async (term: string): Promise<Article[]> => {
  // const response = await fetch(`${BASE_URL}search?query=${encodeURIComponent(term)}`, {
  //   method: 'GET',
  //   headers: {
  //     'Content-Type': 'application/json',
  //   },
  // });
  // if (!response.ok) {
  //   const data = await response.json();
  //   throw new Error(data.error);
  // }
  // const data = await response.json();
  // return data.articles;
  // return []
};

export const fetchNewsByCategory = async (
  category: string | null,
  page: number = 1,
  pageSize: number = 16
): Promise<Article[]> => {
  // const categorySearch = category === 'Home' || '' ? '' : category;
  // const response = await fetch(
  //   `${BASE_URL}latest/?category=${categorySearch}&page=${page}&pageSize=${pageSize}`
  // );
  // if (!response.ok) {
  //   const data = await response.json();
  //   throw new Error(data.error);
  // }
  // const data = await response.json();
  // return data.articles;
  // return []
};
