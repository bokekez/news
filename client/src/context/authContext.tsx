import { createContext, useState, ReactNode, FC } from 'react';
import { User, AuthContextType } from '../types/userModels';
import { showToastifySuccess } from '../config/toastifyConfig';

const AuthContext = createContext<AuthContextType | undefined>(undefined);

export const AuthProvider: FC<{ children: ReactNode }> = ({ children }) => {
  const [user, setUser] = useState<User | null>(null);

  const login = (userData: User) => {
    setUser(userData);
    showToastifySuccess('Logged in successfully!');
  };

  const logout = () => {
    setUser(null);
    showToastifySuccess('Logged out successfully!');
  };

  return (
    <AuthContext.Provider value={{ user, setUser, login, logout }}>
      {children}
    </AuthContext.Provider>
  );
};

export { AuthContext };