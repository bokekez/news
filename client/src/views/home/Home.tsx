import React, { useEffect, useState, useContext } from 'react';
import { fetchNewsBySearchTerm } from '../../api/newsApi';
import { fetchFavorites } from '../../api/favoritesApi';
import styles from './Home.module.scss';
import { ArticleCards } from '../../types/articleModel';
import TopBar from '../../components/TopBar/TopBar';
import Spinner from '../../components/Spinner/Spinner';
import NewsCard from '../../components/NewsCard/NewsCard';
import Categories from '../../components/Categories/Categories';
import { fetchNewsByCategory } from '../../api/newsApi';
import NewsWidget from '../../components/NewsWidget/NewsWidget';
import { showToastifyError } from '../../config/toastifyConfig';
import LoginDialog from '../../components/LoginDialog/LoginDialog';
import { AuthContext } from '../../context/authContext';
import HomeHeader from '../../components/HomeHeader/HomeHeader';
import Button from '../../components/Button/Button';

const Home: React.FC = () => {
  const [searchTerm, setSearchTerm] = useState('');
  const [news, setNews] = useState<ArticleCards[]>([]);
  const [activeCategory, setActiveCategory] = useState<string>('Home');
  const [loading, setLoading] = useState<boolean>(true);
  const [loginDialog, setLoginDialog] = useState<boolean>(false);
  const [showHeader, setShowHeader] = useState<boolean>(true);
  const [isSmallScreen, setIsSmallScreen] = useState<boolean>(false);
  const [smallFeed, setSmallFeed] = useState<string>('Featured');
  const [showSmallCat, setShowSmallCat] = useState<boolean>(false);
  const authContext = useContext(AuthContext);

  const handleSearch = async () => {
    setLoading(true);
    setSmallFeed('Featured');
    setShowSmallCat(false);
  };

  const handleCategorySelect = (category: string) => {
    setSearchTerm('');
    setActiveCategory(category);
  };

  const fetchOrSearchNews = () => {
    if (searchTerm) return fetchNewsBySearchTerm(searchTerm);
    if (activeCategory !== 'Favorites') return fetchNewsByCategory(activeCategory);
    if (authContext?.user?.id) return fetchFavorites(authContext.user.id);
  };

  useEffect(() => {
    if (window.innerWidth <= 768) setIsSmallScreen(true);
    const getNews = async () => {
      try {
        const fetchedNews = await fetchOrSearchNews();
        setNews((fetchedNews as ArticleCards[]) || []);
      } catch (error) {
        showToastifyError(String(error));
      } finally {
        setLoading(false);
      }
    };

    getNews();
  }, [activeCategory, loading]);

  const goToLogin = () => setLoginDialog(true);

  if (!isSmallScreen && news.length <= 16) news.splice(2, 0, { widget: true });

  return (
    <div>
      {showHeader && !isSmallScreen && <HomeHeader setShowHeader={setShowHeader} />}
      <div className={styles.homeContainer}>
        <TopBar
          searchTerm={searchTerm}
          setSearchTerm={setSearchTerm}
          onSearch={handleSearch}
          onLogin={goToLogin}
          isSmallScreen={isSmallScreen}
          setShowSmallCat={setShowSmallCat}
          showSmallCat={showSmallCat}
        />
        <div className={styles.dividerBar} />
        <div className={styles.newsContent}>
          {!isSmallScreen ||
            (showSmallCat && (
              <Categories onCategorySelect={handleCategorySelect} activeCategory={activeCategory} />
            ))}
          {!showSmallCat && (
            <div className={styles.cardsAndWidgetContainer}>
              {isSmallScreen && (
                <div className={styles.smallButtons}>
                  <Button
                    label="Featured"
                    onClick={() => setSmallFeed('Featured')}
                    variant={`${smallFeed === 'Featured' ? 'smallActive' : 'smallInactive'}`}
                  />
                  <Button
                    label="Latest"
                    onClick={() => setSmallFeed('Latest')}
                    variant={`${smallFeed === 'Latest' ? 'smallActive' : 'smallInactive'}`}
                  />
                </div>
              )}
              {!isSmallScreen && <span className={styles.newsHeader}>News</span>}
              {isSmallScreen && smallFeed === 'Latest' && <NewsWidget category="general" />}
              {(!isSmallScreen || (isSmallScreen && smallFeed === 'Featured')) && (
                <div className={styles.cardsContainer}>
                  {news.length > 0 ? (
                    news.map((item, index) =>
                      item.widget ? (
                        <NewsWidget key={`widget-${index}`} category="general" />
                      ) : (
                        <NewsCard
                          key={index}
                          article={item}
                          setLoading={setLoading}
                          category={activeCategory}
                        />
                      )
                    )
                  ) : (
                    <Spinner />
                  )}
                </div>
              )}
            </div>
          )}
        </div>
        {loginDialog && <LoginDialog handleClose={() => setLoginDialog(false)} />}
      </div>
    </div>
  );
};

export default Home;
