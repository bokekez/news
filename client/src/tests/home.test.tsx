import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import { fetchNewsBySearchTerm, fetchNewsByCategory } from '../api/newsApi';
import Home from '../views/home/Home';
import { AuthContext } from '../context/authContext';
import { Article } from '../types/articleModel';

jest.mock('../api/newsApi', () => ({
  fetchNewsBySearchTerm: jest.fn(),
  fetchNewsByCategory: jest.fn(),
}));

jest.mock('../api/favoritesApi', () => ({
  fetchFavorites: jest.fn(),
}));

jest.mock('../components/NewsCard/NewsCard', () => ({
  __esModule: true,
  default: (props: { article: Article }) => <div>{props.article.title || 'News Card'}</div>,
}));

jest.mock('../components/NewsWidget/NewsWidget', () => ({
  __esModule: true,
  default: () => <div>News Widget</div>,
}));

jest.mock('../context/authContext', () => ({
  AuthContext: React.createContext({
    user: { id: '123', name: 'Test User' },
  }),
}));

jest.mock('../config/toastifyConfig', () => ({
  showToastifyError: jest.fn(),
}));

jest.mock('../components/LoginDialog/LoginDialog', () => ({
  __esModule: true,
  default: (props: { handleClose: () => void }) => (
    <button onClick={props.handleClose}>Close Dialog</button>
  ),
}));

jest.mock('../components/Button/Button', () => ({
  __esModule: true,
  default: (props: { label: string; onClick: () => void }) => (
    <button onClick={props.onClick}>{props.label}</button>
  ),
}));

// jest.mock('../components/TopBar/TopBar', () => ({
//   __esModule: true,
//   default: ({ searchTerm, setSearchTerm, setShowSmallCat }: any) => (
//     <div>
//       <input
//         placeholder="Search"
//         value={searchTerm}
//         onChange={(e) => setSearchTerm(e.target.value)}
//       />
//       <button onClick={() => setShowSmallCat(true)}>Toggle Categories</button>
//     </div>
//   ),
// }));

// jest.mock('../components/Categories/Categories', () => ({
//   __esModule: true,
//   default: ({ onCategorySelect, activeCategory }: { onCategorySelect: (category: string) => void, activeCategory: string }) => (
//     <div data-testid="categories-mock">
//       <button onClick={() => onCategorySelect('Technology')} data-testid="category-technology">
//         Technology
//       </button>
//       <p>Active Category: {activeCategory}</p>
//     </div>
//   ),
// }));

const mockArticle: Article = {
  source: {
    id: 'BBC',
    name: 'British Broadcasting Corporation',
  },
  author: 'Test Name',
  title: 'Test Title',
  description: 'Random description',
  url: 'www.bbc.com',
  urlToImage: 'www.bbc.com',
  publishedAt: '09-12-2024',
  content: 'News here',
};

const mockUser = { id: '1', username: 'Test User', firstName: 'test', lastName: 'user' };

const mockSetUser = jest.fn();
const mockLogin = jest.fn();
const mockLogout = jest.fn();

describe('Home Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('renders Home component', () => {
    render(
      <AuthContext.Provider
        value={{ user: mockUser, setUser: mockSetUser, login: mockLogin, logout: mockLogout }}
      >
        <Home />
      </AuthContext.Provider>
    );
    expect(screen.getByText(/Make News your homepage/i)).toBeInTheDocument();
  });

  test('fetches news by search term', async () => {
    const mockFetchNewsBySearchTerm = fetchNewsBySearchTerm as jest.MockedFunction<
      typeof fetchNewsBySearchTerm
    >;
    mockFetchNewsBySearchTerm.mockResolvedValue([mockArticle]);
    render(
      <AuthContext.Provider
        value={{ user: mockUser, setUser: mockSetUser, login: mockLogin, logout: mockLogout }}
      >
        <Home />
      </AuthContext.Provider>
    );

    const searchInput = screen.getByPlaceholderText(/Search news/i);
    const searchButton = screen.getByText(/Search/i);

    fireEvent.change(searchInput, { target: { value: 'Technology' } });
    fireEvent.click(searchButton);

    await waitFor(() => {
      expect(mockFetchNewsBySearchTerm).toHaveBeenCalledWith('Technology');
    });
  });

  test('fetches news by category when selected', async () => {
    const mockFetchNewsByCategory = fetchNewsByCategory as jest.MockedFunction<
      typeof fetchNewsByCategory
    >;
    mockFetchNewsByCategory.mockResolvedValue([mockArticle]);
    render(
      <AuthContext.Provider
        value={{ user: mockUser, setUser: mockSetUser, login: mockLogin, logout: mockLogout }}
      >
        <Home />
      </AuthContext.Provider>
    );

    fireEvent.click(screen.getByText(/Health/i));

    await waitFor(() => {
      expect(mockFetchNewsByCategory).toHaveBeenCalledWith('Health');
    });
  });

  test('renders mobile layout for small screens', () => {
    global.innerWidth = 500;
    global.dispatchEvent(new Event('resize'));
    render(
      <AuthContext.Provider
        value={{ user: mockUser, setUser: mockSetUser, login: mockLogin, logout: mockLogout }}
      >
        <Home />
      </AuthContext.Provider>
    );

    expect(screen.getByText(/Featured/i)).toBeInTheDocument();
    expect(screen.getByText(/Latest/i)).toBeInTheDocument();
  });

  test('closes login dialog on click', () => {
    const noUser = { id: '', username: '', firstName: '', lastName: '' };
    render(
      <AuthContext.Provider
        value={{ user: noUser, setUser: mockSetUser, login: mockLogin, logout: mockLogout }}
      >
        <Home />
      </AuthContext.Provider>
    );
    fireEvent.click(screen.getByText(/Login/i));
    fireEvent.click(screen.getByText(/Close Dialog/i));
    expect(screen.queryByText(/Close Dialog/i)).not.toBeInTheDocument();
  });

  test('renders Categories component on small screen when showSmallCat is true', async () => {
    global.innerWidth = 500;
    global.dispatchEvent(new Event('resize'));

    render(
      <AuthContext.Provider
        value={{ user: mockUser, setUser: mockSetUser, login: mockLogin, logout: mockLogout }}
      >
        <Home />
      </AuthContext.Provider>
    );

    const toggleButton = screen.getByTestId('toggle-small-categories');
    fireEvent.click(toggleButton);

    await waitFor(() => expect(screen.getByText('Sports')).toBeInTheDocument());
  });

  test('matches snapshot', () => {
    global.innerWidth = 500;
    global.dispatchEvent(new Event('resize'));
    const { asFragment } = render(
      <AuthContext.Provider
        value={{ user: mockUser, setUser: mockSetUser, login: mockLogin, logout: mockLogout }}
      >
        <Home />
      </AuthContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
