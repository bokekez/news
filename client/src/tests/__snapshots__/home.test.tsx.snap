// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="homeContainer"
    >
      <div
        class="topBar"
      >
        <span
          class="headline"
        >
          <span
            class="my"
          >
            My
          </span>
          <span
            class="news"
          >
            News
          </span>
        </span>
        <div
          class="searchContainer"
        >
          <img
            src="test-file-stub"
          />
          <input
            class="searchInput"
            placeholder="Search news"
            type="text"
            value=""
          />
        </div>
        <div
          class="userMenu"
        >
          <button>
            Test User
          </button>
        </div>
        <img
          class="smallCategories"
          data-testid="toggle-small-categories"
          src="test-file-stub"
        />
      </div>
      <div
        class="dividerBar"
      />
      <div
        class="newsContent"
      >
        <div
          class="cardsAndWidgetContainer"
        >
          <div
            class="smallButtons"
          >
            <button>
              Featured
            </button>
            <button>
              Latest
            </button>
          </div>
          <div
            class="cardsContainer"
          >
            <div>
              News Widget
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
